File: meddraft_ai/prompts/medical_research_skills/scientific-skills/Data Analysis/meta-funnel-plot/scripts/funnel_plot.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Funnel plot drawing and publication bias test script (Python version)
Usage: python funnel_plot.py <csv_path> <type> [outcome_name] [output_dir]"""

import sys
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
import warnings

warnings.filterwarnings('ignore')

# Set Chinese font
plt.rcParams['font.sans-serif'] = ['SimHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False

def parse_args():
    """Parse command line parameters"""
    if len(sys.argv) < 3:
        print("Usage: python funnel_plot.py <csv_path> <type> [outcome_name] [output_dir]")
        sys.exit(1)
    
    csv_path = sys.argv[1]
    data_type = sys.argv[2]
    outcome_name = sys.argv[3] if len(sys.argv) > 3 and sys.argv[3] else "Outcome"
    output_dir = sys.argv[4] if len(sys.argv) > 4 and sys.argv[4] else os.path.dirname(csv_path)
    
    if data_type not in ["Binary", "Continuity", "Survival"]:
        print("type must be Binary, Continuity or Survival")
        sys.exit(1)
    
    return csv_path, data_type, outcome_name, output_dir

def read_data(csv_path):
    """Read CSV data"""
    try:
        df = pd.read_csv(csv_path)
        return df
    except Exception as e:
        print(f"Failed to read file: {e}")
        sys.exit(1)

def calculate_binary_effect(group1_events, group1_n, group2_events, group2_n):
    """Calculate odds ratios and standard errors for binary data"""
    group1_n_events = group1_n - group1_events
    group2_n_events = group2_n - group2_events
    
    # Calculate OR and 95%CI
    or_value = (group1_events * group2_n_events) / (group1_n_events * group2_events)
    se_log_or = np.sqrt(1/group1_events + 1/group1_n_events + 1/group2_events + 1/group2_n_events)
    
    log_or = np.log(or_value)
    se = se_log_or
    
    return log_or, se, or_value

def calculate_continuity_effect(n1, mean1, sd1, n2, mean2, sd2):
    """Calculate the standardized mean difference for continuous data"""
    pooled_sd = np.sqrt(((n1-1)*sd1**2 + (n2-1)*sd2**2) / (n1 + n2 - 2))
    smd = (mean1 - mean2) / pooled_sd
    se = np.sqrt(1/n1 + 1/n2 + smd**2 / (2*(n1+n2)))
    
    return smd, se

def calculate_survival_effect(hr, ci_lower, ci_upper):
    """Processing survival data"""
    log_hr = np.log(hr)
    se_log_hr = (np.log(ci_upper) - np.log(ci_lower)) / (2 * 1.96)
    
    return log_hr, se_log_hr

def egger_test(log_effects, ses):
    """Egger linear regression test"""
    precision = 1 / np.array(ses)
    
    # linear regression
    x = precision
    y = np.array(log_effects) * precision
    
    n = len(x)
    x_mean = np.mean(x)
    y_mean = np.mean(y)
    
    slope = np.sum((x - x_mean) * (y - y_mean)) / np.sum((x - x_mean)**2)
    intercept = y_mean - slope * x_mean
    
    # Calculate residuals and standard errors
    y_pred = intercept + slope * x
    residuals = y - y_pred
    mse = np.sum(residuals**2) / (n - 2)
    se_intercept = np.sqrt(mse * (1/n + x_mean**2 / np.sum((x - x_mean)**2)))
    
    # t-test
    t_stat = intercept / se_intercept
    p_value = 2 * (1 - stats.t.cdf(abs(t_stat), n - 2))
    
    return intercept, se_intercept, t_stat, p_value

def begg_test(log_effects, ses):
    """Begg rank correlation test"""
    try:
        from scipy.stats import kendalltau
        
        # Calculate Kendall's tau
        tau, p_value = kendalltau(log_effects, ses)
        
        # Calculate z value
        n = len(log_effects)
        z_stat = tau * np.sqrt(2 * (2*n + 5) / (9*n*(n-1)))
        
        return tau, z_stat, p_value
    except:
        return 0, 0, 1.0

def trim_and_fill(log_effects, ses):
    """Cut and complement analysis"""
    n = len(log_effects)
    sorted_effects = np.sort(log_effects)
    
    # A simple estimate of the number of imputed studies
    if n > 2:
        center = np.median(log_effects)
        filled_studies = sum(1 for e in log_effects if e < center) - sum(1 for e in log_effects if e > center)
        filled_studies = max(0, filled_studies)
    else:
        filled_studies = 0
    
    # Calculate pooled effect size (simple weighted average)
    weights = 1 / np.array(ses)**2
    pooled_effect_before = np.sum(np.array(log_effects) * weights) / np.sum(weights)
    
    if filled_studies > 0:
        # Add filled studies
        augmented_effects = list(log_effects) + [np.median(log_effects)] * filled_studies
        augmented_ses = list(ses) + [np.median(ses)] * filled_studies
        weights_after = 1 / np.array(augmented_ses)**2
        pooled_effect_after = np.sum(np.array(augmented_effects) * weights_after) / np.sum(weights_after)
    else:
        pooled_effect_after = pooled_effect_before
    
    return pooled_effect_before, pooled_effect_after, filled_studies

def plot_funnel(log_effects, ses, outcome_name, output_path, data_type):
    """Draw a funnel chart"""
    fig, ax = plt.subplots(figsize=(10, 8))
    
    # Calculate 95% and 97.5% confidence interval bounds
    x_max = max(1/np.array(ses)) * 1.1
    x = np.linspace(0, x_max, 100)
    y_95 = 1.96 * x
    y_975 = 1.96 * x
    
    # Draw confidence intervals
    ax.fill_between(x, -y_95, y_95, alpha=0.1, color='blue', label='95% CI')
    ax.plot(x, y_95, 'b--', alpha=0.5, linewidth=1)
    ax.plot(x, -y_95, 'b--', alpha=0.5, linewidth=1)
    
    # Draw study points
    colors = ['red' if se > np.median(ses) else 'blue' for se in ses]
    ax.scatter(1/np.array(ses), log_effects, alpha=0.6, s=100, c=colors)
    
    # Draw merged effect lines
    pooled = np.sum(np.array(log_effects) / np.array(ses)**2) / np.sum(1 / np.array(ses)**2)
    ax.axvline(pooled, color='red', linestyle='-', linewidth=2, label='Pooled Effect')
    ax.axhline(0, color='black', linestyle='-', linewidth=0.5)
    
    # Set labels and titles
    ax.set_xlabel('1 / Standard Error (Precision)', fontsize=12)
    ax.set_ylabel('log(Effect Size)', fontsize=12)
    ax.set_title(f'Funnel Plot - {outcome_name}', fontsize=14, fontweight='bold')
    ax.legend(fontsize=10)
    ax.grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(output_path, dpi=300, bbox_inches='tight')
    plt.close()

def main():
    """main program"""
    csv_path, data_type, outcome_name, output_dir = parse_args()
    
    # Verify file exists
    if not os.path.exists(csv_path):
        print(f"File does not exist: {csv_path}")
        sys.exit(1)
    
    # Create output directory
    os.makedirs(output_dir, exist_ok=True)
    
    # Read data
    df = read_data(csv_path)
    
    # Verification data volume
    if len(df) < 2:
        print("At least 2 studies are required to draw a funnel plot")
        sys.exit(1)
    
    # Process data according to type
    log_effects = []
    ses = []
    study_names = df['study'].tolist()
    
    try:
        if data_type == "Binary":
            required_cols = ['group1_Events', 'group1_sample_size', 'group2_Events', 'group2_sample_size']
            missing = [c for c in required_cols if c not in df.columns]
            if missing:
                print(f"Missing required column: {', '.join(missing)}")
                sys.exit(1)
            
            for idx, row in df.iterrows():
                log_or, se, or_value = calculate_binary_effect(
                    row['group1_Events'], row['group1_sample_size'],
                    row['group2_Events'], row['group2_sample_size']
                )
                log_effects.append(log_or)
                ses.append(se)
        
        elif data_type == "Continuity":
            required_cols = ['group1_sample_size', 'group1_Mean', 'group1_SD', 
                           'group2_sample_size', 'group2_Mean', 'group2_SD']
            missing = [c for c in required_cols if c not in df.columns]
            if missing:
                print(f"Missing required column: {', '.join(missing)}")
                sys.exit(1)
            
            for idx, row in df.iterrows():
                smd, se = calculate_continuity_effect(
                    row['group1_sample_size'], row['group1_Mean'], row['group1_SD'],
                    row['group2_sample_size'], row['group2_Mean'], row['group2_SD']
                )
                log_effects.append(smd)
                ses.append(se)
        
        elif data_type == "Survival":
            required_cols = ['group1_HR', 'group1_95%Lower CI', 'group1_95%Upper CI']
            missing = [c for c in required_cols if c not in df.columns]
            if missing:
                print(f"Missing required column: {', '.join(missing)}")
                sys.exit(1)
            
            for idx, row in df.iterrows():
                log_hr, se_log_hr = calculate_survival_effect(
                    row['group1_HR'], row['group1_95%Lower CI'], row['group1_95%Upper CI']
                )
                log_effects.append(log_hr)
                ses.append(se_log_hr)
    
    except Exception as e:
        print(f"Data processing failed: {e}")
        sys.exit(1)
    
    log_effects = np.array(log_effects)
    ses = np.array(ses)
    
    # Perform statistical tests
    egger_intercept, egger_se, egger_t, egger_p = egger_test(log_effects, ses)
    begg_tau, begg_z, begg_p = begg_test(log_effects, ses)
    pooled_before, pooled_after, filled_n = trim_and_fill(log_effects, ses)
    
    # Convert back to original effect size
    if data_type == "Binary":
        or_before = np.exp(pooled_before)
        or_after = np.exp(pooled_after)
        ci_before = (np.exp(pooled_before - 1.96 * np.sqrt(np.sum(1/ses**2))**-1), 
                    np.exp(pooled_before + 1.96 * np.sqrt(np.sum(1/ses**2))**-1))
        ci_after = (np.exp(pooled_after - 1.96 * np.sqrt(np.sum(1/ses**2))**-1),
                   np.exp(pooled_after + 1.96 * np.sqrt(np.sum(1/ses**2))**-1))
    else:
        or_before = or_after = ci_before = ci_after = None
    
    # Save results
    base_name = f"{data_type}_funnel_{outcome_name}"
    
    # Draw a funnel chart
    plot_funnel(log_effects, ses, outcome_name, os.path.join(output_dir, f"{base_name}.png"), data_type)
    
    # Save funnel chart data
    funnel_data = pd.DataFrame({
        'study': study_names,
        'log_effect': log_effects,
        'se': ses,
        'precision': 1/ses
    })
    funnel_data.to_csv(os.path.join(output_dir, f"{base_name}.csv"), index=False)
    
    # Save Egger test results
    egger_data = pd.DataFrame({
        'Test': ['Egger Linear Regression'],
        'Intercept': [egger_intercept],
        'SE': [egger_se],
        't_value': [egger_t],
        'p_value': [egger_p]
    })
    egger_data.to_csv(os.path.join(output_dir, f"{data_type}_Egger_{outcome_name}.csv"), index=False)
    
    # Save Begg test results
    begg_data = pd.DataFrame({
        'Test': ['Begg Rank Correlation'],
        'Kendall_tau': [begg_tau],
        'z_value': [begg_z],
        'p_value': [begg_p]
    })
    begg_data.to_csv(os.path.join(output_dir, f"{data_type}_Begg_{outcome_name}.csv"), index=False)
    
    # Output results
    print("\n" + "═"*50)
    print("Funnel plot drawing and publication bias test completed")
    print("═"*50 + "\n")
    
    print(f"【Outcome indicators】{outcome_name}")
    print(f"【data type】{data_type}")
    print(f"【Included studies】{len(df)} item\n")
    
    print("【Output file】")
    print(f"• funnel chart：{os.path.join(output_dir, f'{base_name}.png')}")
    print(f"• funnel data：{os.path.join(output_dir, f'{base_name}.csv')}")
    print(f"• Eggertest：{os.path.join(output_dir, f'{data_type}_Egger_{outcome_name}.csv')}")
    print(f"• Beggtest：{os.path.join(output_dir, f'{data_type}_Begg_{outcome_name}.csv')}\n")
    
    print("[Publication bias test results]")
    print("Egger linear regression test:")
    print(f"• intercept = {egger_intercept:.4f} (SE = {egger_se:.4f})")
    print(f"• tvalue = {egger_t:.4f}")
    print(f"• Pvalue = {egger_p:.4f}")
    print(f"• in conclusion：{'Significant publication bias' if egger_p < 0.05 else 'No significant publication bias'}\n")
    
    print("Begg rank correlation test:")
    print(f"• Kendall's tau = {begg_tau:.4f}")
    print(f"• zvalue = {begg_z:.4f}")
    print(f"• Pvalue = {begg_p:.4f}")
    print(f"• in conclusion：{'Significant publication bias' if begg_p < 0.05 else 'No significant publication bias'}\n")
    
    if data_type == "Binary" and or_before and or_after:
        print("[Analysis of cut and fill method]")
        print(f"• Before trimming：{or_before:.2f} [{ci_before[0]:.2f}; {ci_before[1]:.2f}]")
        print(f"• After trimming：{or_after:.2f} [{ci_after[0]:.2f}; {ci_after[1]:.2f}]")
        print(f"• Fill in the number of studies：{filled_n}\n")
    
    print("═"*50)

if __name__ == "__main__":
    main()

File: meddraft_ai/prompts/medical_research_skills/scientific-skills/Data Analysis/meta-funnel-plot/scripts/funnel_plot.R
#!/usr/bin/env Rscript
# 漏斗图绘制与发表偏倚检验脚本
# 用法: Rscript funnel_plot.R <csv_path> <type> [outcome_name] [output_dir]

suppressPackageStartupMessages({
  library(meta)
  library(metafor)
  library(stringr)
})

# 解析命令行参数
args <- commandArgs(trailingOnly = TRUE)

if (length(args) < 2) {
  stop("用法: Rscript funnel_plot.R <csv_path> <type> [outcome_name] [output_dir]")
}

csv_path <- args[1]
type <- args[2]
outcome_name <- if (length(args) >= 3 && args[3] != "") args[3] else NULL
output_dir <- if (length(args) >= 4 && args[4] != "") args[4] else dirname(csv_path)

# 验证类型
if (!type %in% c("Binary", "Continuity", "Survival")) {
  stop("type 必须是 Binary, Continuity 或 Survival")
}

# 读取数据
if (!file.exists(csv_path)) {
  stop(paste0("文件不存在: ", csv_path))
}

data_df <- read.csv(csv_path, stringsAsFactors = FALSE)

# 获取结局指标名称
if (is.null(outcome_name)) {
  if ("outcome_new" %in% colnames(data_df)) {
    outcome_name <- unique(data_df$outcome_new)[1]
  } else {
    outcome_name <- "Outcome"
  }
}

# 验证数据量
if (nrow(data_df) < 2) {
  stop("至少需要2项研究才能绘制漏斗图")
}

# 创建输出目录
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

# 辅助函数：获取列
get_col <- function(df, candidates) {
  for (col in candidates) {
    if (col %in% colnames(df)) return(df[[col]])
  }
  return(NULL)
}

# 根据类型创建meta分析对象
if (type == "Binary") {
  required_cols <- c("study", "group1_Events", "group1_sample_size",
                     "group2_Events", "group2_sample_size")
  missing_cols <- setdiff(required_cols, colnames(data_df))
  if (length(missing_cols) > 0) {
    stop(paste0("缺少必要列: ", paste(missing_cols, collapse = ", ")))
  }

  meta_analysis <- metabin(
    event.e = as.numeric(data_df$group1_Events),
    n.e = as.numeric(data_df$group1_sample_size),
    event.c = as.numeric(data_df$group2_Events),
    n.c = as.numeric(data_df$group2_sample_size),
    studlab = data_df$study,
    sm = "OR",
    method = "MH",
    random = TRUE
  )
  effect_name <- "OR"

} else if (type == "Continuity") {
  required_cols <- c("study", "group1_sample_size", "group1_Mean", "group1_SD",
                     "group2_sample_size", "group2_Mean", "group2_SD")
  missing_cols <- setdiff(required_cols, colnames(data_df))
  if (length(missing_cols) > 0) {
    stop(paste0("缺少必要列: ", paste(missing_cols, collapse = ", ")))
  }

  meta_analysis <- metacont(
    studlab = data_df$study,
    n.e = as.numeric(data_df$group1_sample_size),
    mean.e = as.numeric(data_df$group1_Mean),
    sd.e = as.numeric(data_df$group1_SD),
    n.c = as.numeric(data_df$group2_sample_size),
    mean.c = as.numeric(data_df$group2_Mean),
    sd.c = as.numeric(data_df$group2_SD),
    sm = "SMD",
    method.tau = "DL",
    method.random.ci = "HK"
  )
  effect_name <- "SMD"

} else if (type == "Survival") {
  hr_col <- get_col(data_df, c("group1_HR"))
  lower_col <- get_col(data_df, c("group1_95%Lower CI", "group1_95.Lower.CI"))
  upper_col <- get_col(data_df, c("group1_95%Upper CI", "group1_95.Upper.CI"))

  if (is.null(hr_col) || is.null(lower_col) || is.null(upper_col)) {
    stop("缺少必要列: group1_HR, group1_95%Lower CI, group1_95%Upper CI")
  }

  hr <- as.numeric(hr_col)
  lower_ci <- as.numeric(lower_col)
  upper_ci <- as.numeric(upper_col)

  valid_idx <- !is.na(hr) & !is.na(lower_ci) & !is.na(upper_ci) &
    hr > 0 & lower_ci > 0 & upper_ci > 0

  if (sum(valid_idx) < 2) {
    stop("有效数据不足2项，无法绘制漏斗图")
  }

  log_hr <- log(hr[valid_idx])
  se <- (log(upper_ci[valid_idx]) - log(lower_ci[valid_idx])) / (2 * 1.96)

  meta_analysis <- metagen(
    TE = log_hr,
    seTE = se,
    studlab = data_df$study[valid_idx],
    sm = "HR",
    method.tau = "DL",
    common = TRUE
  )
  effect_name <- "HR"
}

# 绘制漏斗图
funnel_file <- file.path(output_dir, paste0(type, "_funnel_", outcome_name, ".png"))

png(funnel_file, width = 800, height = 800, units = 'px', pointsize = 20)
test <- funnel(meta_analysis, lty.random = 0)
dev.off()

# 保存漏斗图数据
funnel_data <- as.data.frame(t(rbind(test$xvals, test$yvals)))
colnames(funnel_data) <- c('x', 'y')
funnel_csv <- file.path(output_dir, paste0(type, "_funnel_", outcome_name, ".csv"))
write.csv(funnel_data, funnel_csv, row.names = FALSE)

# 发表偏倚检验（需要至少3项研究）
n_studies <- length(meta_analysis$studlab)
Egger_result <- NULL
Begg_result <- NULL
Effect_size_tf <- NULL

if (n_studies >= 3) {
  # Egger线性回归检验
  Egger_res <- tryCatch({
    metabias(meta_analysis, k.min = 2, method.bias = "Egger")
  }, error = function(e) NULL)

  if (!is.null(Egger_res)) {
    Egger_result <- data.frame(
      beta = Egger_res$estimate[1],
      se.beta = Egger_res$estimate[2],
      intercept = Egger_res$intercept,
      se.intercept = Egger_res$se.intercept,
      statistic = Egger_res$statistic,
      p_val = Egger_res$pval,
      df = Egger_res$df
    )
    egger_csv <- file.path(output_dir, paste0(type, "_Egger_", outcome_name, ".csv"))
    write.csv(Egger_result, egger_csv, row.names = FALSE)
  }

  # Begg秩相关检验
  Begg_res <- tryCatch({
    metabias(meta_analysis, k.min = 2, method.bias = "Begg")
  }, error = function(e) NULL)

  if (!is.null(Begg_res)) {
    Begg_result <- data.frame(
      ks = Begg_res$estimate[1],
      se.ks = Begg_res$estimate[2],
      statistic = Begg_res$statistic,
      p.val = Begg_res$pval
    )
    begg_csv <- file.path(output_dir, paste0(type, "_Begg_", outcome_name, ".csv"))
    write.csv(Begg_result, begg_csv, row.names = FALSE)
  }

  # 剪补法（Trim and Fill）
  tf1 <- tryCatch({
    meta::trimfill(meta_analysis, ma.common = FALSE, common = FALSE, random = TRUE)
  }, error = function(e) NULL)

  if (!is.null(tf1)) {
    n_filled <- sum(tf1$trimfill)
    if (n_filled > 0) {
      Effect_size_tf <- data.frame(
        method = c("Trim Fill before", "Trim Fill after"),
        effect = c(round(exp(meta_analysis$TE.random), 2), round(exp(tf1$TE.random), 2)),
        lower = c(round(exp(meta_analysis$lower.random), 2), round(exp(tf1$lower.random), 2)),
        upper = c(round(exp(meta_analysis$upper.random), 2), round(exp(tf1$upper.random), 2)),
        n_filled = c(0, n_filled)
      )
      tf_csv <- file.path(output_dir, paste0(type, "_Trimfill_", outcome_name, ".csv"))
      write.csv(Effect_size_tf, tf_csv, row.names = FALSE)
    }
  }
}

# 输出结果摘要
cat("\n")
cat("═══════════════════════════════════════════\n")
cat("漏斗图绘制与发表偏倚检验完成\n")
cat("═══════════════════════════════════════════\n\n")
cat(paste0("【结局指标】", outcome_name, "\n"))
cat(paste0("【数据类型】", type, "\n"))
cat(paste0("【纳入研究】", n_studies, " 项\n\n"))
cat("【输出文件】\n")
cat(paste0("• 漏斗图：", funnel_file, "\n"))
cat(paste0("• 漏斗数据：", funnel_csv, "\n"))

if (!is.null(Egger_result)) {
  cat(paste0("• Egger检验：", file.path(output_dir, paste0(type, "_Egger_", outcome_name, ".csv")), "\n"))
}
if (!is.null(Begg_result)) {
  cat(paste0("• Begg检验：", file.path(output_dir, paste0(type, "_Begg_", outcome_name, ".csv")), "\n"))
}

cat("\n【发表偏倚检验结果】\n\n")

if (!is.null(Egger_result)) {
  egger_conclusion <- if (Egger_result$p_val < 0.05) "存在显著发表偏倚" else "无显著发表偏倚"
  cat("Egger线性回归检验：\n")
  cat(paste0("• 截距 = ", round(Egger_result$intercept, 4),
             " (SE = ", round(Egger_result$se.intercept, 4), ")\n"))
  cat(paste0("• t值 = ", round(Egger_result$statistic, 4), "\n"))
  cat(paste0("• P值 = ", round(Egger_result$p_val, 4), "\n"))
  cat(paste0("• 结论：", egger_conclusion, "\n\n"))
} else {
  cat("Egger检验：研究数量不足，无法进行检验\n\n")
}

if (!is.null(Begg_result)) {
  begg_conclusion <- if (Begg_result$p.val < 0.05) "存在显著发表偏倚" else "无显著发表偏倚"
  cat("Begg秩相关检验：\n")
  cat(paste0("• Kendall's tau = ", round(Begg_result$ks, 4), "\n"))
  cat(paste0("• z值 = ", round(Begg_result$statistic, 4), "\n"))
  cat(paste0("• P值 = ", round(Begg_result$p.val, 4), "\n"))
  cat(paste0("• 结论：", begg_conclusion, "\n\n"))
} else {
  cat("Begg检验：研究数量不足，无法进行检验\n\n")
}

if (!is.null(Effect_size_tf)) {
  cat("【剪补法分析】\n")
  cat(paste0("• 剪补前：", Effect_size_tf$effect[1],
             " [", Effect_size_tf$lower[1], "; ", Effect_size_tf$upper[1], "]\n"))
  cat(paste0("• 剪补后：", Effect_size_tf$effect[2],
             " [", Effect_size_tf$lower[2], "; ", Effect_size_tf$upper[2], "]\n"))
  cat(paste0("• 填补研究数：", Effect_size_tf$n_filled[2], "\n"))
}

cat("═══════════════════════════════════════════\n")
